ml_perf/reference_implementation.py

```diff
@@ -0,0 +1,279 @@
+
+  @property
+  def play_model_path(self):
+    return os.path.join(fsdb.models_dir(), self.play_model_name)
+
+  @property
+  def train_model_name(self):
+    return '%06d-%s' % (self.iter_num, self._NAMES[self.train_model_num])
+
+  @property
+  def train_model_path(self):
+    return os.path.join(fsdb.models_dir(), self.train_model_name)
+
+  @property
+  def seed(self):
+    return self.iter_num + 1
+
+
+def checked_run(cmd, name):
+  logging.info('Running %s:\n  %s', name, '\n  '.join(cmd))
+  with utils.logged_timer('%s finished' % name.capitalize()):
+    completed_process = subprocess.run(
+        cmd, stdout=subprocess.PIPE, stderr=subprocess.STDOUT)
+    if completed_process.returncode:
+      logging.error('Error running %s: %s', name,
+                    completed_process.stdout.decode())
+      raise RuntimeError('Non-zero return code executing %s' % ' '.join(cmd))
+  return completed_process
+
+
+def get_lines(completed_process, slice):
+  return '\n'.join(completed_process.stdout.decode()[:-1].split('\n')[slice])
+
+
+class MakeSlice(object):
+
+  def __getitem__(self, item):
+    return item
+
+
+make_slice = MakeSlice()
+
+
+def cc_flags(state):
+  return [
+      '--engine={}'.format(FLAGS.engine),
+      '--virtual_losses=8',
+      '--seed={}'.format(state.seed),
+  ]
+
+
+def py_flags(state):
+  return [
+      '--work_dir={}'.format(fsdb.working_dir()),
+      '--trunk_layers=10',
+      '--conv_width=64',
+      '--value_cost_weight=0.25',
+      '--training_seed={}'.format(state.seed),
+  ]
+
+
+# Generate an initial model with random weights.
+def bootstrap(state):
+  checked_run([
+      'python3', 'bootstrap.py', '--export_path={}'.format(
+          state.play_model_path)
+  ] + py_flags(state), 'bootstrap')
+
+
+# Freezes a model checkpoint into a TensorFlow GraphDef proto, which is the
+# required model format for C++ TensorFlow.
+def freeze(state, model_path):
+  checked_run([
+      'python3', 'freeze_graph.py', '--model_path={}'.format(model_path)
+  ] + py_flags(state), 'freeze')
+
+
+# Self-play a number of games.
+def selfplay(state):
+  play_output_name = state.play_output_name
+  play_output_dir = os.path.join(fsdb.selfplay_dir(), play_output_name)
+  play_holdout_dir = os.path.join(fsdb.holdout_dir(), play_output_name)
+
+  result = checked_run([
+      'bazel-bin/cc/selfplay', '--parallel_games=2048',
+      '--num_readouts=100', '--model={}.pb'.format(
+          state.play_model_path), '--output_dir={}'.format(play_output_dir),
+      '--holdout_dir={}'.format(play_holdout_dir)
+  ] + cc_flags(state), 'selfplay')
+  logging.info(get_lines(result, make_slice[-2:]))
+
+  # Write examples to a single record.
+  pattern = os.path.join(play_output_dir, '*', '*.zz')
+  logging.info('Extracting examples from "{}"'.format(pattern))
+  random.seed(state.seed)
+  tensorflow.set_random_seed(state.seed)
+  numpy.random.seed(state.seed)
+  buffer = example_buffer.ExampleBuffer(sampling_frac=1.0)
```

Comment:
wow, this code lives on. I'm slightly horrified. It's also on the main path, so its speed actually matters too.

Reply:
Another thing I disliked about the examplebuffer was that it was set up for in-process instead of subprocessed. if this whole thing is going to be run continuously in one python process, the memory usage of the EB could be less than ideal.

Reply:
added a TODO